File: Axiom/Internal/Intermediate/ImMemoryRead.cs
using System.IO;

namespace Axiom.Internal.Intermediate
{
    internal class ImMemoryRead : ImExpression
    {
        public ImMemoryRead(ImExpression expression)
        {
            Expression = expression;
        }

        public override void Dump(TextWriter tw, int ident)
        {
            Report.Dump(tw, ident, "MemoryRead");
            Expression.Dump(tw, ident + 1);
        }

        public override ImExpressionSequence Linear()
        {
            var lin = Expression.Linear();

            lin.Expression = new ImMemoryRead(lin.Expression);

            return lin;
        }

        public ImExpression Expression { get; set; }
    }
}

File: src/main/java/com/moybl/axiom/im/ImMemoryRead.java
package com.moybl.axiom.im;

public class ImMemoryRead implements ImExpression {

	private ImExpression expression;

	public ImMemoryRead(ImExpression expression) {
		this.expression = expression;
	}

	@Override
	public void accept(ImVisitor visitor) {
		visitor.visit(this);
	}

	public ImExpression getExpression() {
		return expression;
	}

}
